MAPIE is not able to one-hot encode columns when estimator is an scikit learn pipeline with a preprocessor step

**Describe the bug**
Dear colleagues, I am creating a system to classify customers in 2 binary classes and then apply a regression model to one of the classes.

Some of my features are string that I obviously need to encode. In this case with one hot encoding.

**To Reproduce**

My code is as follows:

```python
from sklearnex import patch_sklearn
patch_sklearn()

from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer, TransformedTargetRegressor, make_column_selector as selector 
from sklearn.preprocessing import StandardScaler, OneHotEncoder, LabelEncoder 
from sklearn.impute import SimpleImputer
#from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.multioutput import RegressorChain, MultiOutputRegressor
from sklearn.feature_selection import VarianceThreshold
from sklearn.metrics import mean_absolute_error, make_scorer, mean_tweedie_deviance, auc
from sklearn.model_selection import RandomizedSearchCV, train_test_split, LeaveOneGroupOut, LeavePGroupsOut, cross_validate
from sklearn.metrics import roc_auc_score, plot_roc_curve, roc_curve, confusion_matrix, classification_report, ConfusionMatrixDisplay
from sklearn.ensemble import RandomForestClassifier, HistGradientBoostingRegressor, HistGradientBoostingClassifier

from sklearn import set_config
from mapie.regression import MapieRegressor

data_train, data_test, target_train, target_test = train_test_split(
    df.drop(columns=target_reg + target_class + METADATA_COLUMNS), 
    df[target_reg + target_class], 
    random_state=42)

categorical_columns_of_interest = ['col1', 'col2', 'col3', 'col4', 'col5', 'col6', 'col7']
numerical_columns = ml_data.drop(columns=target_reg + target_class + METADATA_COLUMNS).select_dtypes(include=np.number).columns
numerical_columns = [x for x in MY_FEATURES if x not in FEATURES_NOT_TO_IMPUTE]
numerical_columns = [x for x in numerical_columns if x not in categorical_columns_of_interest]

categorical_transformer = OneHotEncoder(handle_unknown="ignore")
numeric_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="mean")), 
        ("scaler", StandardScaler()),
        ("variance_selector", VarianceThreshold(threshold=0.03))
        ]
)
preprocessor = ColumnTransformer(
    transformers=[
        ("numeric_only", numeric_transformer, numerical_columns),
        ("get_dummies", categorical_transformer, categorical_columns_of_interest)])

pipeline_hist_boost_reg= Pipeline([('preprocessor', preprocessor),
                             ('estimator', HistGradientBoostingRegressor())])

regressor = TransformedTargetRegressor(pipeline_hist_boost_reg, func=np.log1p, inverse_func=np.expm1)

mapie_estimator = MapieRegressor(pipeline_hist_boost_reg)
mapie_estimator.fit(data_train, target_train)
```

**Expected behavior**
After this I will expect that I can run:

y_pred, y_pis = mapie_estimator.predict(data_test)

**Screenshots**


```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-43-61f70ee71787> in <module>
      1 mapie_estimator = MapieRegressor(pipeline_hist_boost_reg)
----> 2 mapie_estimator.fit(X_train_reg, y_train_reg)

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/mapie/regression.py in fit(self, X, y, sample_weight)
    457         cv = self._check_cv(self.cv)
    458         estimator = self._check_estimator(self.estimator)
--> 459         X, y = check_X_y(
    460             X, y, force_all_finite=False, dtype=["float64", "int", "object"]
    461         )

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/sklearn/utils/validation.py in check_X_y(X, y, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, estimator)
    962         raise ValueError("y cannot be None")
    963 
--> 964     X = check_array(
    965         X,
    966         accept_sparse=accept_sparse,

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/sklearn/utils/validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator)
    683     if has_pd_integer_array:
    684         # If there are any pandas integer extension arrays,
--> 685         array = array.astype(dtype)
    686 
    687     if force_all_finite not in (True, False, "allow-nan"):

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/generic.py in astype(self, dtype, copy, errors)
   5804         else:
   5805             # else, only a single dtype is given
-> 5806             new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
   5807             return self._constructor(new_data).__finalize__(self, method="astype")
   5808 

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/internals/managers.py in astype(self, dtype, copy, errors)
    412 
    413     def astype(self: T, dtype, copy: bool = False, errors: str = "raise") -> T:
--> 414         return self.apply("astype", dtype=dtype, copy=copy, errors=errors)
    415 
    416     def convert(

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/internals/managers.py in apply(self, f, align_keys, ignore_failures, **kwargs)
    325                     applied = b.apply(f, **kwargs)
    326                 else:
--> 327                     applied = getattr(b, f)(**kwargs)
    328             except (TypeError, NotImplementedError):
    329                 if not ignore_failures:

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/internals/blocks.py in astype(self, dtype, copy, errors)
    590         values = self.values
    591 
--> 592         new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
    593 
    594         new_values = maybe_coerce_values(new_values)

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/dtypes/cast.py in astype_array_safe(values, dtype, copy, errors)
   1298 
   1299     try:
-> 1300         new_values = astype_array(values, dtype, copy=copy)
   1301     except (ValueError, TypeError):
   1302         # e.g. astype_nansafe can fail on object-dtype of strings

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/dtypes/cast.py in astype_array(values, dtype, copy)
   1246 
   1247     else:
-> 1248         values = astype_nansafe(values, dtype, copy=copy)
   1249 
   1250     # in pandas we don't store numpy str dtypes, so convert to object

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/dtypes/cast.py in astype_nansafe(arr, dtype, copy, skipna)
   1083         flags = arr.flags
   1084         flat = arr.ravel("K")
-> 1085         result = astype_nansafe(flat, dtype, copy=copy, skipna=skipna)
   1086         order: Literal["C", "F"] = "F" if flags.f_contiguous else "C"
   1087         # error: Item "ExtensionArray" of "Union[ExtensionArray, ndarray]" has no

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/pandas/core/dtypes/cast.py in astype_nansafe(arr, dtype, copy, skipna)
   1190     if copy or is_object_dtype(arr.dtype) or is_object_dtype(dtype):
   1191         # Explicit copy, or required since NumPy can't view from / to object.
-> 1192         return arr.astype(dtype, copy=True)
   1193 
   1194     return arr.astype(dtype, copy=copy)

ValueError: could not convert string to float: 'group C'
```
Being this value part of one of the categorical columns which its being encoded by the preprocessor. 

When training the model without mapie everything works correctly:

![image](https://user-images.githubusercontent.com/14976422/151325317-6843b79c-a3ee-456b-9e2b-22086ffb2027.png)

**Desktop (please complete the following information):**

```python
import platform
print(platform.machine())
print(platform.version())
print(platform.platform())
print(platform.system())
print(platform.processor())

x86_64
#58~18.04.1-Ubuntu SMP Wed Jul 28 23:14:18 UTC 2021
Linux-5.4.0-1056-azure-x86_64-with-glibc2.10
Linux
x86_64
```
Scikit learn dependencies:

```
scikit-learn==1.0.2
scikit-learn-intelex==2021.5.1
imbalance-learn==0.9
mapie==0.3.1
```

Hi @gmartinonQM unfortunately I can not share the data but. I have tried with another dataset attached here:
[kaggle_dataset.csv](https://github.com/scikit-learn-contrib/MAPIE/files/7951963/kaggle_dataset.csv)

Which I got from this kaggle kernel:

https://www.kaggle.com/riantowibisono/loan-classification-machine-learning/notebook

After doing:

```python
import numpy as nb
import pandas as pd
df = pd.read_csv('kaggle_dataset.csv')
df_1 = df.dropna(axis=1, thresh=int(0.70*len(df)))
df_1.head()
df_clean = df_1[[
    'loan_status_fullyPaid', 'term','int_rate',
    'installment','grade', 'annual_inc',
    'verification_status','dti'  # These features are just initial guess, you can try to choose any other combination
]].copy()
df_clean.head()
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()

df_clean['term'] = label.fit_transform(df_clean['term'])
df_clean['grade'] = label.fit_transform(df_clean['grade'])
df_clean['verification_status'] = label.fit_transform(df_clean['verification_status'])

x = df_clean.drop(['loan_status_fullyPaid'], axis=1)
y = df_clean['loan_status_fullyPaid']

from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.compose import ColumnTransformer, TransformedTargetRegressor, make_column_selector as selector 
from sklearn.feature_selection import VarianceThreshold
from sklearn.preprocessing import StandardScaler, OneHotEncoder, LabelEncoder 
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
import numpy as np 

categorical_transformer = OneHotEncoder(handle_unknown="ignore", categories='auto')
numeric_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="mean")), 
        ("scaler", StandardScaler()),
        ("variance_selector", VarianceThreshold(threshold=0.03))
        ]
)
preprocessor = ColumnTransformer(
    transformers=[
        ("numeric_only", numeric_transformer, [2,4,6]),
        ("get_dummies", categorical_transformer, [0,3,5])], 
        remainder = 'passthrough'                               
)

from sklearn.model_selection import train_test_split
xtr, xts, ytr, yts = train_test_split(
    x,
    y,
    test_size = .2
)

from sklearn.metrics import roc_auc_score, plot_roc_curve, roc_curve, confusion_matrix, classification_report, ConfusionMatrixDisplay
from sklearn.ensemble import RandomForestClassifier, HistGradientBoostingRegressor, HistGradientBoostingClassifier

pipeline_hist_boost_clf = Pipeline([('preprocessor', preprocessor),
                             ('estimator', HistGradientBoostingClassifier())])

pipeline_hist_boost_clf.fit(xtr, ytr)

from mapie.classification import MapieClassifier

mapie_classifier = MapieClassifier(pipeline_hist_boost_clf)
mapie_classifier.fit(xtr, ytr)
```

This time with a classification model following the example of: 

https://mapie.readthedocs.io/en/latest/examples_classification/plot_sadinle2019_example.html

I get a similar error:

```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-27-b25355019e0d> in <module>
      2 
      3 mapie_classifier = MapieClassifier(pipeline_hist_boost_clf)
----> 4 mapie_classifier.fit(xtr, ytr)

/anaconda/envs/azureml_py38/lib/python3.8/site-packages/mapie/classification.py in fit(self, X, y, sample_weight)
    511             X, y, force_all_finite=False, dtype=["float64", "int", "object"]
    512         )
--> 513         assert type_of_target(y) == "multiclass"
    514         self.n_features_in_ = check_n_features_in(X, cv, estimator)
    515         sample_weight, X, y = check_null_weight(sample_weight, X, y)

AssertionError: 

```

It seems that this **force_all_finite** is the main issue. I will be happy to contribute to the debugging, if needed I could offer myself for a teams call.

BR
E

--- a/.github/PULL_REQUEST_TEMPLATE.md
+++ b/.github/PULL_REQUEST_TEMPLATE.md
@@ -20,7 +20,7 @@ Please describe the tests that you ran to verify your changes. Provide instructi
 - [ ] Test A
 - [ ] Test B
 
-# Checklist:
+# Checklist
 
 - [ ] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
 - [ ] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -14,6 +14,7 @@ History
 "predict" in regression.py
 * Add replication of the Chen Xu's tutorial testing Jackknife+aB vs Jackknife+
 * Add Jackknife+-after-Bootstrap documentation
+* Improve scikit-learn pipelines compatibility for regression
 
 0.3.1 (2021-11-19)
 ------------------

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -14,7 +14,7 @@ History
 "predict" in regression.py
 * Add replication of the Chen Xu's tutorial testing Jackknife+aB vs Jackknife+
 * Add Jackknife+-after-Bootstrap documentation
-* Improve scikit-learn pipelines compatibility for regression
+* Improve scikit-learn pipelines compatibility
 
 0.3.1 (2021-11-19)
 ------------------

--- a/environment.dev.yml
+++ b/environment.dev.yml
@@ -5,7 +5,7 @@ channels:
 dependencies:
     - bump2version=1.0.1
     - flake8=4.0.1
-    - mypy=0.920
+    - mypy=0.910
     - numpydoc=1.1.0
     - pandas=1.3.5
     - pytest=6.2.5

--- a/examples/classification/1-quickstart/plot_comp_methods_on_2d_dataset.py
+++ b/examples/classification/1-quickstart/plot_comp_methods_on_2d_dataset.py
@@ -168,7 +168,7 @@ def plot_scores(
 fig, axs = plt.subplots(1, 2, figsize=(10, 5))
 for i, method in enumerate(methods):
     conformity_scores = mapie[method].conformity_scores_
-    n = mapie[method].n_samples_val_
+    n = mapie[method].n_samples_
     quantiles = mapie[method].quantiles_
     plot_scores(alpha, conformity_scores, quantiles, method, axs[i])
 plt.show()

--- a/mapie/classification.py
+++ b/mapie/classification.py
@@ -7,10 +7,15 @@ from sklearn.base import BaseEstimator, ClassifierMixin, clone
 from sklearn.linear_model import LogisticRegression
 from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import label_binarize
 from sklearn.utils import check_random_state, _safe_indexing
 from sklearn.utils.multiclass import type_of_target
-from sklearn.utils.validation import indexable, check_is_fitted, _num_samples, _check_y
-from sklearn.preprocessing import label_binarize
+from sklearn.utils.validation import (
+    indexable,
+    check_is_fitted,
+    _num_samples,
+    _check_y,
+)
 
 from ._typing import ArrayLike
 from ._machine_precision import EPSILON

--- a/mapie/classification.py
+++ b/mapie/classification.py
@@ -9,7 +9,7 @@ from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
 from sklearn.utils import check_random_state, _safe_indexing
 from sklearn.utils.multiclass import type_of_target
-from sklearn.utils.validation import check_is_fitted
+from sklearn.utils.validation import indexable, check_is_fitted, _num_samples
 from sklearn.preprocessing import label_binarize
 
 from ._typing import ArrayLike
@@ -122,7 +122,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
     n_features_in_: int
         Number of features passed to the fit method.
 
-    n_samples_val_: Union[int, List[int]]
+    n_samples_: Union[int, List[int]]
         Number of samples passed to the fit method.
 
     conformity_scores_ : ArrayLike of shape (n_samples_train)
@@ -173,7 +173,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
         "single_estimator_",
         "estimators_",
         "n_features_in_",
-        "n_samples_val_",
+        "n_samples_",
         "conformity_scores_"
     ]
 
@@ -612,7 +612,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
             estimator = fit_estimator(
                 estimator, X_train, y_train, sample_weight[train_index]
             )
-        if X_val.shape[0] > 0:
+        if _num_samples(X_val) > 0:
             y_pred_proba = self._predict_oof_model(
                 estimator, X_val,
             )
@@ -665,13 +665,9 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
         self._check_parameters()
         cv = check_cv(self.cv)
         estimator = self._check_estimator(X, y, self.estimator)
-
         if self.image_input:
             check_input_is_image(X)
-        # X, y = check_X_y(
-        #     X, y, force_all_finite=False, ensure_2d=self.image_input,
-        #     allow_nd=self.image_input, dtype=["float64", "int", "object"]
-        # )
+        X, y = indexable(X, y)
         assert type_of_target(y) == "multiclass"
         self.n_classes_ = len(set(y))
         self.n_features_in_ = check_n_features_in(X, cv, estimator)
@@ -680,10 +676,9 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
         # Initialization
         self.estimators_: List[ClassifierMixin] = []
         self.k_ = np.empty_like(y, dtype=int)
-        self.n_samples_val_ = X.shape[0]
+        self.n_samples_ = _num_samples(X)
 
         # Work
-
         if cv == "prefit":
             self.single_estimator_ = estimator
             y_pred_proba = self.single_estimator_.predict_proba(X)
@@ -832,13 +827,6 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
         check_is_fitted(self, self.fit_attributes)
         if self.image_input:
             check_input_is_image(X)
-        # X = check_array(
-        #     X,
-        #     force_all_finite=False,
-        #     ensure_2d=self.image_input,
-        #     allow_nd=self.image_input,
-        #     dtype=["float64", "object"]
-        # )
 
         # Estimate probabilities from estimator(s)
         y_pred = self.single_estimator_.predict(X)
@@ -868,7 +856,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
                 raise ValueError("Invalid 'agg_scores' argument.")
 
         # Estimate prediction sets
-        n = self.n_samples_val_
+        n = self.n_samples_
         if alpha_ is None:
             return np.array(y_pred)
 

--- a/mapie/classification.py
+++ b/mapie/classification.py
@@ -9,7 +9,7 @@ from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
 from sklearn.utils import check_random_state, _safe_indexing
 from sklearn.utils.multiclass import type_of_target
-from sklearn.utils.validation import indexable, check_is_fitted, _num_samples
+from sklearn.utils.validation import indexable, check_is_fitted, _num_samples, _check_y
 from sklearn.preprocessing import label_binarize
 
 from ._typing import ArrayLike
@@ -668,6 +668,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
         if self.image_input:
             check_input_is_image(X)
         X, y = indexable(X, y)
+        y = _check_y(y)
         assert type_of_target(y) == "multiclass"
         self.n_classes_ = len(set(y))
         self.n_features_in_ = check_n_features_in(X, cv, estimator)

--- a/mapie/classification.py
+++ b/mapie/classification.py
@@ -7,7 +7,7 @@ from sklearn.base import BaseEstimator, ClassifierMixin, clone
 from sklearn.linear_model import LogisticRegression
 from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
-from sklearn.utils import check_X_y, check_array, check_random_state
+from sklearn.utils import check_random_state, _safe_indexing
 from sklearn.utils.multiclass import type_of_target
 from sklearn.utils.validation import check_is_fitted
 from sklearn.preprocessing import label_binarize
@@ -601,9 +601,11 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
           of shape (n_samples_val,).
 
         """
-        X_train, y_train, X_val, y_val = (
-            X[train_index], y[train_index], X[val_index], y[val_index]
-        )
+        X_train = _safe_indexing(X, train_index)
+        y_train = _safe_indexing(y, train_index)
+        X_val = _safe_indexing(X, val_index)
+        y_val = _safe_indexing(y, val_index)
+
         if sample_weight is None:
             estimator = fit_estimator(estimator, X_train, y_train)
         else:
@@ -666,10 +668,10 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
 
         if self.image_input:
             check_input_is_image(X)
-        X, y = check_X_y(
-            X, y, force_all_finite=False, ensure_2d=self.image_input,
-            allow_nd=self.image_input, dtype=["float64", "int", "object"]
-        )
+        # X, y = check_X_y(
+        #     X, y, force_all_finite=False, ensure_2d=self.image_input,
+        #     allow_nd=self.image_input, dtype=["float64", "int", "object"]
+        # )
         assert type_of_target(y) == "multiclass"
         self.n_classes_ = len(set(y))
         self.n_features_in_ = check_n_features_in(X, cv, estimator)
@@ -681,6 +683,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
         self.n_samples_val_ = X.shape[0]
 
         # Work
+
         if cv == "prefit":
             self.single_estimator_ = estimator
             y_pred_proba = self.single_estimator_.predict_proba(X)
@@ -716,7 +719,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
             self.conformity_scores_ = np.empty(y_pred_proba.shape)
         elif self.method == "score":
             self.conformity_scores_ = np.take_along_axis(
-                1 - y_pred_proba, y.reshape(-1, 1), axis=1
+                1 - y_pred_proba, np.ravel(y).reshape(-1, 1), axis=1
             )
         elif self.method == "cumulated_score":
             y_true = label_binarize(y=y, classes=estimator.classes_)
@@ -731,7 +734,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
                 y_pred_proba_sorted_cumsum, cutoff.reshape(-1, 1), axis=1
             )
             y_proba_true = np.take_along_axis(
-                y_pred_proba, y.reshape(-1, 1), axis=1
+                y_pred_proba, np.ravel(y).reshape(-1, 1), axis=1
             )
             random_state = check_random_state(self.random_state)
             u = random_state.uniform(size=len(y_pred_proba)).reshape(-1, 1)
@@ -744,7 +747,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
             )
             self.conformity_scores_ = np.take_along_axis(
                 index,
-                y.reshape(-1, 1),
+                np.ravel(y).reshape(-1, 1),
                 axis=1
             )
 
@@ -829,13 +832,13 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):  # type: ignore
         check_is_fitted(self, self.fit_attributes)
         if self.image_input:
             check_input_is_image(X)
-        X = check_array(
-            X,
-            force_all_finite=False,
-            ensure_2d=self.image_input,
-            allow_nd=self.image_input,
-            dtype=["float64", "object"]
-        )
+        # X = check_array(
+        #     X,
+        #     force_all_finite=False,
+        #     ensure_2d=self.image_input,
+        #     allow_nd=self.image_input,
+        #     dtype=["float64", "object"]
+        # )
 
         # Estimate probabilities from estimator(s)
         y_pred = self.single_estimator_.predict(X)

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -9,7 +9,7 @@ from sklearn.base import BaseEstimator, RegressorMixin, clone
 from sklearn.linear_model import LinearRegression
 from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
-from sklearn.utils import check_array, _safe_indexing
+from sklearn.utils import _safe_indexing
 from sklearn.utils.validation import check_is_fitted
 
 from ._typing import ArrayLike

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -346,7 +346,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         val_index: ArrayLike,
         k: int,
         sample_weight: Optional[ArrayLike] = None,
-    ) -> Tuple[RegressorMixin, ArrayLike, ArrayLike, ArrayLike]:
+    ) -> Tuple[RegressorMixin, ArrayLike, ArrayLike]:
         """
         Fit a single out-of-fold model on a given training set and
         perform predictions on a test set.
@@ -377,12 +377,10 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
 
         Returns
         -------
-        Tuple[RegressorMixin, ArrayLike, ArrayLike, ArrayLike]
+        Tuple[RegressorMixin, ArrayLike, ArrayLike]
 
         - [0]: Fitted estimator
         - [1]: Estimator predictions on the validation fold,
-          of shape (n_samples_val,)
-        - [2]: Identification number of the validation fold,
           of shape (n_samples_val,)
         - [3]: Validation data indices,
           of shape (n_samples_val,).

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -541,7 +541,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
                 ]
 
                 for i, val_ind in enumerate(val_indices):
-                    pred_matrix[val_ind, i] = np.ravel(predictions[i])
+                    pred_matrix[val_ind, i] = predictions[i].ravel()
                     self.k_[val_ind, i] = 1
                 check_nan_in_aposteriori_prediction(pred_matrix)
 

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -604,7 +604,6 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         check_is_fitted(self, self.fit_attributes)
         self._check_ensemble(ensemble)
         alpha_ = check_alpha(alpha)
-        X = check_array(X, force_all_finite=False, dtype=["float64", "object"])
         y_pred = self.single_estimator_.predict(X)
 
         if alpha is None:
@@ -657,7 +656,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
                         )
                         for _alpha in alpha_
                     ]
-                )
+                ).data
                 y_pred_up = np.column_stack(
                     [
                         np.quantile(
@@ -668,7 +667,8 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
                         )
                         for _alpha in alpha_
                     ]
-                )
+                ).data
                 if ensemble:
                     y_pred = aggregate_all(self.agg_function, y_pred_multi)
+                np.stack([y_pred_low, y_pred_up], axis=1)
             return y_pred, np.stack([y_pred_low, y_pred_up], axis=1)

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -10,7 +10,12 @@ from sklearn.linear_model import LinearRegression
 from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
 from sklearn.utils import _safe_indexing
-from sklearn.utils.validation import indexable, check_is_fitted, _num_samples, _check_y
+from sklearn.utils.validation import (
+    indexable,
+    check_is_fitted,
+    _num_samples,
+    _check_y,
+)
 
 from ._typing import ArrayLike
 from .aggregation_functions import aggregate_all, phi2D

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -541,7 +541,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
                 ]
 
                 for i, val_ind in enumerate(val_indices):
-                    pred_matrix[val_ind, i] = predictions[i].ravel()
+                    pred_matrix[val_ind, i] = np.array(predictions[i]).ravel()
                     self.k_[val_ind, i] = 1
                 check_nan_in_aposteriori_prediction(pred_matrix)
 

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -10,7 +10,7 @@ from sklearn.linear_model import LinearRegression
 from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
 from sklearn.utils import _safe_indexing
-from sklearn.utils.validation import check_is_fitted
+from sklearn.utils.validation import indexable, check_is_fitted, _num_samples
 
 from ._typing import ArrayLike
 from .aggregation_functions import aggregate_all, phi2D
@@ -149,7 +149,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
     n_features_in_: int
         Number of features passed to the fit method.
 
-    n_samples_val_: List[int]
+    n_samples_: List[int]
         Number of samples passed to the fit method.
 
     References
@@ -190,7 +190,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         "k_",
         "residuals_",
         "n_features_in_",
-        "n_samples_val_",
+        "n_samples_",
     ]
 
     def __init__(
@@ -393,7 +393,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
             estimator = fit_estimator(
                 estimator, X_train, y_train, sample_weight_train
             )
-        if X_val.shape[0] > 0:
+        if _num_samples(X_val) > 0:
             y_pred = estimator.predict(X_val)
         else:
             y_pred = np.array([])
@@ -481,6 +481,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         cv = check_cv(self.cv)
         estimator = self._check_estimator(self.estimator)
         agg_function = self._check_agg_function(self.agg_function)
+        X, y = indexable(X, y)
         self.n_features_in_ = check_n_features_in(X, cv, estimator)
         sample_weight, X, y = check_null_weight(sample_weight, X, y)
 
@@ -491,7 +492,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         if cv == "prefit":
             self.single_estimator_ = estimator
             y_pred = self.single_estimator_.predict(X)
-            self.n_samples_val_ = [X.shape[0]]
+            self.n_samples_ = [_num_samples(X)]
             self.k_ = np.full(
                 shape=(len(y), 1), fill_value=np.nan, dtype=float
             )
@@ -513,7 +514,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
             )
             if self.method == "naive":
                 y_pred = self.single_estimator_.predict(X)
-                self.n_samples_val_ = [X.shape[0]]
+                self.n_samples_ = [_num_samples(X)]
             else:
                 outputs = Parallel(n_jobs=self.n_jobs, verbose=self.verbose)(
                     delayed(self._fit_and_predict_oof_model)(
@@ -531,7 +532,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
                     list, zip(*outputs)
                 )
 
-                self.n_samples_val_ = [
+                self.n_samples_ = [
                     np.array(pred).shape[0] for pred in predictions
                 ]
 

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -9,7 +9,7 @@ from sklearn.base import BaseEstimator, RegressorMixin, clone
 from sklearn.linear_model import LinearRegression
 from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
-from sklearn.utils import check_array, check_X_y
+from sklearn.utils import check_array, _safe_indexing
 from sklearn.utils.validation import check_is_fitted
 
 from ._typing import ArrayLike
@@ -383,19 +383,21 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
           of shape (n_samples_val,).
 
         """
-        X_train, y_train, X_val = X[train_index], y[train_index], X[val_index]
+        X_train = _safe_indexing(X, train_index)
+        y_train = _safe_indexing(y, train_index)
+        X_val = _safe_indexing(X, val_index)
         if sample_weight is None:
             estimator = fit_estimator(estimator, X_train, y_train)
         else:
+            sample_weight_train = _safe_indexing(sample_weight, train_index)
             estimator = fit_estimator(
-                estimator, X_train, y_train, sample_weight[train_index]
+                estimator, X_train, y_train, sample_weight_train
             )
         if X_val.shape[0] > 0:
             y_pred = estimator.predict(X_val)
         else:
             y_pred = np.array([])
-        val_id = np.full_like(y_pred, k, dtype=int)
-        return estimator, y_pred, val_id, val_index
+        return estimator, y_pred, val_index
 
     def aggregate_with_mask(self, x: ArrayLike, k: ArrayLike) -> ArrayLike:
         """
@@ -479,9 +481,6 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         cv = check_cv(self.cv)
         estimator = self._check_estimator(self.estimator)
         agg_function = self._check_agg_function(self.agg_function)
-        X, y = check_X_y(
-            X, y, force_all_finite=False, dtype=["float64", "int", "object"]
-        )
         self.n_features_in_ = check_n_features_in(X, cv, estimator)
         sample_weight, X, y = check_null_weight(sample_weight, X, y)
 
@@ -528,7 +527,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
                     )
                     for k, (train_index, val_index) in enumerate(cv.split(X))
                 )
-                self.estimators_, predictions, val_ids, val_indices = map(
+                self.estimators_, predictions, val_indices = map(
                     list, zip(*outputs)
                 )
 
@@ -537,13 +536,13 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
                 ]
 
                 for i, val_ind in enumerate(val_indices):
-                    pred_matrix[val_ind, i] = predictions[i]
+                    pred_matrix[val_ind, i] = np.ravel(predictions[i])
                     self.k_[val_ind, i] = 1
                 check_nan_in_aposteriori_prediction(pred_matrix)
 
                 y_pred = aggregate_all(agg_function, pred_matrix)
 
-        self.residuals_ = np.abs(y - y_pred)
+        self.residuals_ = np.abs(np.ravel(y) - y_pred)
         return self
 
     def predict(

--- a/mapie/regression.py
+++ b/mapie/regression.py
@@ -10,7 +10,7 @@ from sklearn.linear_model import LinearRegression
 from sklearn.model_selection import BaseCrossValidator
 from sklearn.pipeline import Pipeline
 from sklearn.utils import _safe_indexing
-from sklearn.utils.validation import indexable, check_is_fitted, _num_samples
+from sklearn.utils.validation import indexable, check_is_fitted, _num_samples, _check_y
 
 from ._typing import ArrayLike
 from .aggregation_functions import aggregate_all, phi2D
@@ -482,6 +482,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         estimator = self._check_estimator(self.estimator)
         agg_function = self._check_agg_function(self.agg_function)
         X, y = indexable(X, y)
+        y = _check_y(y)
         self.n_features_in_ = check_n_features_in(X, cv, estimator)
         sample_weight, X, y = check_null_weight(sample_weight, X, y)
 

--- a/mapie/tests/test_classification.py
+++ b/mapie/tests/test_classification.py
@@ -347,11 +347,6 @@ class WrongOutputModel:
         ).astype(int)
         return pred
 
-    def get_params(self, deep: bool = False) -> Dict[str, ArrayLike]:
-        return {
-            "proba_out": self.proba_out
-        }
-
 
 def do_nothing(*args: Any) -> None:
     "Mock function that does nothing."

--- a/mapie/tests/test_classification.py
+++ b/mapie/tests/test_classification.py
@@ -16,7 +16,6 @@ from sklearn.pipeline import make_pipeline, Pipeline
 from sklearn.impute import SimpleImputer
 from sklearn.preprocessing import OneHotEncoder
 from sklearn.compose import ColumnTransformer
-from sklearn.tree import DecisionTreeClassifier
 from sklearn.utils.validation import check_is_fitted
 
 from mapie.classification import MapieClassifier
@@ -861,7 +860,7 @@ def test_pipeline_compatibility(strategy: str) -> None:
             ("num", numeric_preprocessor, ["x_num"])
         ]
     )
-    pipe = make_pipeline(preprocessor, DecisionTreeClassifier())
+    pipe = make_pipeline(preprocessor, LogisticRegression())
     pipe.fit(X, y)
     mapie = MapieClassifier(estimator=pipe, **STRATEGIES[strategy][0])
     mapie.fit(X, y)

--- a/mapie/tests/test_classification.py
+++ b/mapie/tests/test_classification.py
@@ -3,6 +3,7 @@ from __future__ import annotations
 from typing import Any, Optional, Tuple, Union, Iterable, Dict
 from typing_extensions import TypedDict
 
+import pandas as pd
 import pytest
 import numpy as np
 from sklearn.base import ClassifierMixin
@@ -12,6 +13,10 @@ from sklearn.linear_model import LogisticRegression
 from sklearn.model_selection import KFold, LeaveOneOut
 from sklearn.naive_bayes import GaussianNB
 from sklearn.pipeline import make_pipeline, Pipeline
+from sklearn.impute import SimpleImputer
+from sklearn.preprocessing import OneHotEncoder
+from sklearn.compose import ColumnTransformer
+from sklearn.tree import DecisionTreeClassifier
 from sklearn.utils.validation import check_is_fitted
 
 from mapie.classification import MapieClassifier
@@ -833,3 +838,36 @@ def test_pred_loof_isnan() -> None:
         k=0,
     )
     assert len(y_pred) == 0
+
+
+@pytest.mark.parametrize("strategy", [*STRATEGIES])
+def test_pipeline_compatibility(strategy: str) -> None:
+    """Check that MAPIE works on pipeline based on pandas dataframes"""
+    X = pd.DataFrame(
+        {
+            "x_cat": ["A", "A", "B", "A", "A", "B"],
+            "x_num": [0, 1, 1, 4, np.nan, 5],
+        }
+    )
+    y = pd.Series([0, 1, 2, 0, 1, 0])
+    numeric_preprocessor = Pipeline(
+        [
+            ("imputer", SimpleImputer(strategy="mean")),
+        ]
+    )
+    categorical_preprocessor = Pipeline(
+        steps=[
+            ("encoding", OneHotEncoder(handle_unknown="ignore"))
+        ]
+    )
+    preprocessor = ColumnTransformer(
+        [
+            ("cat", categorical_preprocessor, ["x_cat"]),
+            ("num", numeric_preprocessor, ["x_num"])
+        ]
+    )
+    pipe = make_pipeline(preprocessor, DecisionTreeClassifier())
+    pipe.fit(X, y)
+    mapie = MapieClassifier(estimator=pipe, **STRATEGIES[strategy][0])
+    mapie.fit(X, y)
+    mapie.predict(X)

--- a/mapie/tests/test_regression.py
+++ b/mapie/tests/test_regression.py
@@ -460,7 +460,7 @@ def test_pipeline_compatibility() -> None:
         ]
     )
     categorical_preprocessor = Pipeline(
-        steps = [
+        steps=[
             ("encoding", OneHotEncoder(handle_unknown="ignore"))
         ]
     )

--- a/mapie/tests/test_regression.py
+++ b/mapie/tests/test_regression.py
@@ -3,13 +3,18 @@ from __future__ import annotations
 from itertools import combinations
 from typing import Any, List, Optional, Tuple, Union
 
-import numpy as np
 import pytest
+import numpy as np
+import pandas as pd
 from sklearn.datasets import make_regression
 from sklearn.dummy import DummyRegressor
 from sklearn.linear_model import LinearRegression
 from sklearn.model_selection import KFold, LeaveOneOut, train_test_split
 from sklearn.utils.validation import check_is_fitted
+from sklearn.pipeline import make_pipeline, Pipeline
+from sklearn.impute import SimpleImputer
+from sklearn.preprocessing import OneHotEncoder
+from sklearn.compose import ColumnTransformer
 from typing_extensions import TypedDict
 
 from mapie._typing import ArrayLike
@@ -279,20 +284,16 @@ def test_prediction_between_low_up(strategy: str) -> None:
 
 
 @pytest.mark.parametrize("method", ["plus", "minmax"])
-@pytest.mark.parametrize("cv", [-1, 2, 3, 5])
 @pytest.mark.parametrize("agg_function", ["mean", "median"])
-@pytest.mark.parametrize("alpha", [0.05, 0.1, 0.2])
-def test_prediction_agg_function(
-    method: str, cv: Union[LeaveOneOut, KFold], agg_function: str, alpha: int
-) -> None:
+def test_prediction_agg_function(method: str, agg_function: str) -> None:
     """
     Test that predictions differ when ensemble is True/False,
     but not prediction intervals.
     """
-    mapie = MapieRegressor(method=method, cv=cv, agg_function=agg_function)
+    mapie = MapieRegressor(method=method, cv=2, agg_function=agg_function)
     mapie.fit(X, y)
-    y_pred_1, y_pis_1 = mapie.predict(X, ensemble=True, alpha=alpha)
-    y_pred_2, y_pis_2 = mapie.predict(X, ensemble=False, alpha=alpha)
+    y_pred_1, y_pis_1 = mapie.predict(X, ensemble=True, alpha=0.1)
+    y_pred_2, y_pis_2 = mapie.predict(X, ensemble=False, alpha=0.1)
     np.testing.assert_allclose(y_pis_1[:, 0, 0], y_pis_2[:, 0, 0])
     np.testing.assert_allclose(y_pis_1[:, 1, 0], y_pis_2[:, 1, 0])
     with pytest.raises(AssertionError):
@@ -432,7 +433,7 @@ def test_aggregate_with_mask_with_prefit() -> None:
 def test_pred_loof_isnan() -> None:
     """Test that if validation set is empty then prediction is empty."""
     mapie_reg = MapieRegressor()
-    _, y_pred, _, _ = mapie_reg._fit_and_predict_oof_model(
+    _, y_pred, _ = mapie_reg._fit_and_predict_oof_model(
         estimator=LinearRegression(),
         X=X_toy,
         y=y_toy,
@@ -441,3 +442,35 @@ def test_pred_loof_isnan() -> None:
         k=0,
     )
     assert len(y_pred) == 0
+
+
+def test_pipeline_compatibility() -> None:
+    """Check that MAPIE works on pipeline based on pandas dataframes"""
+    X = pd.DataFrame(
+        {
+            "x_cat": ["A", "A", "B", "A", "A", "B"],
+            "x_num": [0, 1, 1, 4, np.nan, 5],
+            "y": [5, 7, 3, 9, 10, 8]
+        }
+    )
+    y = pd.Series([5, 7, 3, 9, 10, 8])
+    numeric_preprocessor = Pipeline(
+        [
+            ("imputer", SimpleImputer(strategy="mean")),
+        ]
+    )
+    categorical_preprocessor = Pipeline(
+        steps = [
+            ("encoding", OneHotEncoder(handle_unknown="ignore"))
+        ]
+    )
+    preprocessor = ColumnTransformer(
+        [
+            ("cat", categorical_preprocessor, ["x_cat"]),
+            ("num", numeric_preprocessor, ["x_num"])
+        ]
+    )
+    pipe = make_pipeline(preprocessor, LinearRegression())
+    mapie = MapieRegressor(pipe)
+    mapie.fit(X, y)
+    mapie.predict(X)

--- a/mapie/tests/test_regression.py
+++ b/mapie/tests/test_regression.py
@@ -433,6 +433,7 @@ def test_aggregate_with_mask_with_prefit() -> None:
 def test_pred_loof_isnan() -> None:
     """Test that if validation set is empty then prediction is empty."""
     mapie_reg = MapieRegressor()
+    y_pred: ArrayLike
     _, y_pred, _ = mapie_reg._fit_and_predict_oof_model(
         estimator=LinearRegression(),
         X=X_toy,

--- a/mapie/utils.py
+++ b/mapie/utils.py
@@ -398,8 +398,8 @@ def check_input_is_image(X: ArrayLike) -> None:
     Parameters
     ----------
     X: Union[
-        ArrayLike[n_samples, width, height],
-        ArrayLike[n_samples, width, height, n_channels]
+        ArrayLike[n_samples, width, height],
+        ArrayLike[n_samples, width, height, n_channels]
     ]
         Image input
 

--- a/mapie/utils.py
+++ b/mapie/utils.py
@@ -5,7 +5,8 @@ from typing import Any, Iterable, Optional, Tuple, Union, cast
 import numpy as np
 from sklearn.base import ClassifierMixin, RegressorMixin
 from sklearn.model_selection import BaseCrossValidator, KFold, LeaveOneOut
-from sklearn.utils.validation import _check_sample_weight
+from sklearn.utils.validation import _check_sample_weight, _num_features
+from sklearn.utils import _safe_indexing
 
 from ._typing import ArrayLike
 
@@ -58,7 +59,8 @@ def check_null_weight(
     if sample_weight is not None:
         sample_weight = _check_sample_weight(sample_weight, X)
         non_null_weight = sample_weight != 0
-        X, y = X[non_null_weight, :], y[non_null_weight]
+        X = _safe_indexing(X, non_null_weight)
+        y = _safe_indexing(y, non_null_weight)
         sample_weight = sample_weight[non_null_weight]
     return sample_weight, X, y
 
@@ -255,7 +257,13 @@ def check_n_features_in(
     >>> print(check_n_features_in(X))
     5
     """
-    n_features_in: int = X.shape[1]
+    if hasattr(X, "shape"):
+        if len(X.shape) <= 1:
+            n_features_in = 1
+        else:
+            n_features_in = X.shape[1]
+    else:
+        n_features_in = _num_features(X)
     if cv == "prefit" and hasattr(estimator, "n_features_in_"):
         if cast(Any, estimator).n_features_in_ != n_features_in:
             raise ValueError(

